system/gatekeeper/lib/traversal.rego

```diff
@@ -0,0 +1,75 @@
+package lib.traversal
+
+### find_pod(obj)
+#
+# Find a Pod object or template  within the given k8s object. For Pods, `obj`
+# itself is returned; but e.g. for Deployments, `.spec.template` is returned.
+#
+# On success, the returned object will have {"found": true} added to it.
+# If no Pod is found, an object with just {"found": false} is returned.
+#
+# Pods that are owned by a ReplicaSet etc. are ignored to avoid
+# double-reporting. (We report on the highest level possible to reduce the
+# number of total violations that are generated; there is no use in generating
+# 30 violations for 30 replicas of the same Deployment.)
+#
+find_pod(obj) = result {
+  # case 1: `obj` is a Pod itself
+  obj.kind == "Pod"
+  owners := [ref.kind | ref := obj.metadata.ownerReferences[_]; ref.kind == __pod_owners[_]]
+  result := __return_pod_unless_ignored(obj, owners)
+}
+find_pod(obj) = result {
+  # case 2: `obj` contains a PodSpec
+  obj.kind == __violation_owners[_]
+  result := object.union(obj.spec.template, { "found": true })
+}
+find_pod(obj) = result {
+  # case 3: neither Pod nor PodSpec
+  obj.kind != "Pod"
+  count([kind | kind := __violation_owners[_]; obj.kind == kind ]) == 0
```

Comment:
I don't fully understand this construct but could it fail and we need another fallback for find_pod?

Reply:
List comprehensions work somewhat similar to rules on the top level, except that they create lists instead of objects. `[ <expr> | <body> ]` creates a list that, for each successful evaluation of `<body>`, contains the result of `<expr>` for the variables declared within the body. So in this case, this reads as: For each `kind := __violation_owners[_]` that fulfils `obj.kind == kind`, the list will contain `kind`.

So there are two cases: Either `obj.kind` is in `__violation_owners`, then the list will have that kind name as its single member and the `count(...) == 0` will not hold (which is okay, since in this situation "case 2" matches instead). Or if we have an unknown value in `obj.kind`, the list will be empty and `count(...) == 0` will hold.